### DISCUSSION TITLE
Cancelling SceneManager.LoadSceneAsync() / .UnloadSceneAsync() that has been converted to UniTask

### DISCUSSION THREAD

**Discussion by HenrysHouses** [Cysharp/UniTask#513] [Q&A]
I've tried reading the documentation, I looked through the discussions here and googled it but i couldn't find an answer to this. So I'll just have to ask it here.

I am making a scene managing solution for my projects so my question is: How does UniTask handle cancellation of scene loading?

for a little context. In normal unity behavior it is impossible to cancel loading and unloading. And in async the async methods, if the scenes are loaded and not enabled the `AsyncOperation` will still be active since the scenes are not enabled and stays on 0.9 completion. Only way to complete the operation is to enable the scenes, and then let the `AsyncOperation` complete. Until the operation is done the scene can not be unloaded. which means it has to become enabled before it can be unloaded.

UniTask can convert these methods and let them be handled with cancellation tokens. But the question still remains, is the rest of the default unity scene handling functioning as before or does UniTask handle the cancelled scene? unloading it if it was supposed to load, or loading it back in if it was unloading?

Either way I'm questioning if cancelling loading is a feature i even need or if it would even be used. But I've seen how useful UniTask can be so i was curious on if it could work.

Thanks!

**Comment by tigran-sargsyan-w** [comment]
`WithCancellation()` does **not** cancel Unity's underlying `AsyncOperation`.

For a normal Unity `AsyncOperation`, UniTask wraps the operation and watches the supplied `CancellationToken`.

When cancellation is detected, the **UniTask** completes as cancelled, but there is no call in the wrapper that aborts the original Unity operation.

So:

```csharp
await SceneManager.LoadSceneAsync(...)
    .WithCancellation(token);
```

means roughly:

```text
stop awaiting this operation through this UniTask
```

not:

```text
abort the Unity scene load and roll everything back
```

The underlying `LoadSceneAsync` / `UnloadSceneAsync` continues according to Unity's normal semantics.

That also means UniTask does not automatically:

- unload a scene whose load was cancelled;
- reload a scene whose unload was cancelled;
- undo a partially progressed scene operation.

For a scene-management system I would therefore treat cancellation as cancellation of the **request/caller**, not cancellation of Unity's scene operation.

If a load becomes obsolete, one possible pattern is to associate each request with a generation/request ID, allow Unity's `AsyncOperation` to reach a valid completed state, and then explicitly unload/clean up the scene if that request is no longer current.

This is especially relevant with `allowSceneActivation = false`, because the underlying Unity operation still follows Unity's normal 0.9/activation behavior even if the UniTask waiting for it has been cancelled.
